enhancements/project-request-limit.md

```diff
@@ -0,0 +1,345 @@
+available within and outside of the cluster via API aggregation of `kube-apiserver`. The group can be reached at 
+`/apis/admission.project.openshift.io/v1/projectrequestlimits` of the `kube-apiserver`, i.e. via the 
+`kubernetes.default.svc` service hostname inside the cluster.
+   
+The `APIService` that represents the webhook server:
+```
+apiVersion: apiregistration.k8s.io/v1
+kind: APIService
+metadata:
+  name: v1.admission.project.openshift.io
+spec:
+  group: admission.project.openshift.io
+  version: v1
+``` 
+
+The webhook server will have the following topology:
+* The webhook is deployed as a `DaemonSet` server.
+* We restrict the webhook server to the master nodes only. Apply appropriate `nodeSelector` and `tolerations` to the 
+  `PodSpec` so that the `Pods` are scheduled on to the master nodes.
+* The webhook server is reachable to the `kube-apiserver` via `kubernetes.default.svc`.
+```
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+webhooks:
+- clientConfig:
+    service:
+      # reach the webhook via the registered aggregated API
+      namespace: default
+      name: kubernetes
+      path: /apis/admission.project.openshift.io/v1/projectrequestlimits
+    caBundle: KUBE_CA
+```
+
+The call chain to the admission webhook:
+```
+`Project` request -> kube-apiserver -> kube aggregator layer inside the kube-apiserver ->
+   openshift-apiserver -> admission layer in `openshift-apiserver` -> `kube-apiserver` -> 
+     `kube-aggregator` layer inside the `kube-apiserver` -> ProjectRequestLimit webhook 
+```
+
+#### The Operator
+The operator will allow a cluster admin to:
+* Enable or disable `ProjectRequestLimit` admission webhook on a cluster.
+* Specify `limits` on `Project` create request(s) that the `ProjectRequestLimit` admission webhook can enforce.
+* Manage other lifecycle aspects of the operand.
+
+The operator will define a CRD to expose its API. The cluster admin will interact with the operator via a custom resource:
+* We treat the `ProjectRequestLimit` admission webhook as a cluster singleton.
+* The CRD will be `cluster-scoped`.
+* The operator will be reconciling a CR named `cluster`, it will ignore other custom resources.
+
+```
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: projectrequestlimits.operator.project.openshift.io
+spec:
+  group: operator.project.openshift.io
+  version: v1
+  scope: Cluster
+  names:
+    plural: projectrequestlimits
+    singular: projectrequestlimit
+    kind: ProjectRequestLimit
+```  
+
+This is how a cluster-admin can enable the `ProjectRequestLimit` admission webhook on a cluster with `limits` to be enforced. 
+```
+apiVersion: operator.project.openshift.io/v1
+kind: ProjectRequestLimitConfig
+metadata:
+  name: cluster
+spec:
+  limits:
+  - selector:
+      level: admin
+  - selector:
+      level: advanced
+    maxProjects: 10
+  - maxProjects: 2
```

Comment:
no dash

Reply:
this is the current design, I added comments to clarify it.

Reply:
ic